File: sdks/node/src/ipc.ts
import * as fs from 'fs';
import * as net from 'net';
import * as os from 'os';
import * as path from 'path';

const PIPE_PATH = '\\\\.\\pipe\\agentshield';

let reqId = 0;
let sessionId: string | undefined;

function socketPath(): string {
  if (process.platform === 'win32') {
    return PIPE_PATH;
  }
  const runtime = process.env.XDG_RUNTIME_DIR ?? '/tmp';
  return path.join(runtime, 'agentshield', 'daemon.sock');
}

export function daemonAvailable(): boolean {
  if (process.platform === 'win32') {
    try {
      fs.accessSync(PIPE_PATH, fs.constants.R_OK | fs.constants.W_OK);
      return true;
    } catch {
      return false;
    }
  }
  return fs.existsSync(socketPath());
}

export async function analyzeViaDaemon(
  command: string,
  cwd?: string
): Promise<Record<string, unknown> | null> {
  const req = {
    id: ++reqId,
    method: 'analyze',
    params: {
      command,
      cwd: cwd ?? process.cwd(),
      session_id: sessionId ?? null,
    },
  };

  try {
    const resp = await callDaemon(req);
    if (resp.error) return null;
    const result = resp.result as Record<string, unknown> | undefined;
    if (result?.session_id && typeof result.session_id === 'string') {
      sessionId = result.session_id;
    }
    return result ?? null;
  } catch {
    return null;
  }
}

async function callDaemon(req: Record<string, unknown>): Promise<{
  result?: unknown;
  error?: { message: string };
}> {
  const payload = `${JSON.stringify(req)}\n`;

  if (process.platform === 'win32') {
    return new Promise((resolve, reject) => {
      const client = net.createConnection(PIPE_PATH, () => {
        client.write(payload);
      });
      let buf = '';
      client.on('data', (chunk) => {
        buf += chunk.toString('utf8');
        if (buf.includes('\n')) {
          client.end();
          resolve(JSON.parse(buf.trim()) as { result?: unknown; error?: { message: string } });
        }
      });
      client.on('error', reject);
    });
  }

  return new Promise((resolve, reject) => {
    const client = net.createConnection(socketPath(), () => {
      client.write(payload);
    });
    let buf = '';
    client.on('data', (chunk) => {
      buf += chunk.toString('utf8');
      if (buf.includes('\n')) {
        client.end();
        resolve(JSON.parse(buf.trim()) as { result?: unknown; error?: { message: string } });
      }
    });
    client.on('error', reject);
  });
}

File: sdks/python/agentshield/ipc.py
"""Daemon IPC client for AgentShield (Unix socket / Windows named pipe)."""

from __future__ import annotations

import json
import os
import sys
from typing import Any
from uuid import UUID

_REQ_ID = 0
_SESSION_ID: UUID | None = None

if sys.platform == "win32":
    _PIPE_PATH = r"\\.\pipe\agentshield"
else:
    _PIPE_PATH = (
        os.environ.get("XDG_RUNTIME_DIR", "/tmp")
        + "/agentshield/daemon.sock"
    )


def _next_id() -> int:
    global _REQ_ID
    _REQ_ID += 1
    return _REQ_ID


def daemon_available() -> bool:
    if sys.platform == "win32":
        try:
            with open(_PIPE_PATH, "r+b", buffering=0):
                return True
        except OSError:
            return False
    return os.path.exists(_PIPE_PATH)


def analyze_via_daemon(command: str, cwd: str | None = None) -> dict[str, Any] | None:
    global _SESSION_ID
    req = {
        "id": _next_id(),
        "method": "analyze",
        "params": {
            "command": command,
            "cwd": cwd,
            "session_id": str(_SESSION_ID) if _SESSION_ID else None,
        },
    }
    try:
        resp = _call_daemon(req)
    except OSError:
        return None

    if resp.get("error"):
        return None
    result = resp.get("result") or {}
    sid = result.get("session_id")
    if sid:
        _SESSION_ID = UUID(str(sid))
    return result


def _call_daemon(req: dict[str, Any]) -> dict[str, Any]:
    payload = (json.dumps(req) + "\n").encode("utf-8")
    if sys.platform == "win32":
        with open(_PIPE_PATH, "r+b", buffering=0) as pipe:
            pipe.write(payload)
            line = _read_line(pipe)
    else:
        import socket

        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
            sock.settimeout(2.0)
            sock.connect(_PIPE_PATH)
            sock.sendall(payload)
            line = _recv_line(sock)
    return json.loads(line.decode("utf-8"))


def _read_line(pipe: Any) -> bytes:
    buf = bytearray()
    while True:
        chunk = pipe.read(1)
        if not chunk or chunk == b"\n":
            break
        buf.extend(chunk)
    return bytes(buf)


def _recv_line(sock: Any) -> bytes:
    buf = bytearray()
    while True:
        chunk = sock.recv(1)
        if not chunk or chunk == b"\n":
            break
        buf.extend(chunk)
    return bytes(buf)